src/compiler/scala/tools/nsc/util/CachedClassPath.scala

```diff
@@ -0,0 +1,268 @@
+            prefetchStatus += 1
+        }
+        lastCacheValidTime = valid
+      }
+      valid
+    }
+  }
+  private sealed trait ClassesCachedClassPath extends CachedClassPath {
+
+    private val classesCache = new CachedLazyMapping(underlying.classes, Nil)
+    override private[nsc] final def classes(inPackage: String) = classesCache(inPackage)
+
+    override def findClassFile(className: String): Option[AbstractFile] = underlying.findClassFile(className)
+
+    override def mappings: List[BaseCachedLazyMapping[_]] = classesCache :: super.mappings
+
+    override protected def stats = s"${super.stats} classes${classesCache.stats}"
+  }
+  private sealed trait SourcesCachedClassPath  extends CachedClassPath{
+
+    private val sourcesCache = new CachedLazyMapping(underlying.sources, Nil)
+    override private[nsc] def sources(inPackage: String) = sourcesCache(inPackage)
+
+    override lazy val asSourcePathString: String = underlying.asSourcePathString
+
+    override def mappings: List[BaseCachedLazyMapping[_]] = sourcesCache :: super.mappings
+
+    override protected def stats = s"${super.stats} sources${sourcesCache.stats}"
+  }
+
+}
+
+/**
+  * a Cached implementation of a mapping. This has to be threadsafe as a classpath may be cached and reused in a JVM
+  * @tparam V
+  */
+abstract class BaseCachedLazyMapping[V <: AnyRef] {
+
+  @volatile protected var complete = false
+  protected val cache: ConcurrentHashMap[String,_]
+
+  protected val hits = new AtomicInteger
+  protected val misses = new AtomicInteger
+  protected val defaulted = new AtomicInteger
+
+  def apply(key:String): V
+  def markComplete(): Unit = {
+    this.complete = true
+  }
+  def optimise()
+  def stats = s"H$hits M$misses D$defaulted"
+  def reset() :Unit = {
+    cache.clear()
+    complete = false
+  }
+}
+class CachedLazyMapping[V <: AnyRef] (miss : (String => V), defaultValue: V ) extends BaseCachedLazyMapping[V] {
+  override protected val cache =  new ConcurrentHashMap[String,V]
+
+  def apply(key:String): V = {
+    //note - order dependent. Must be before we check the cache
+    val alreadyComplete = complete
+    val cached = cache.get(key)
+    if (cached ne null) {
+      hits.incrementAndGet()
+      cached
+    } else if (alreadyComplete) {
+      defaulted.incrementAndGet()
+      defaultValue
+    } else {
+      misses.incrementAndGet()
+      val missed = miss(key)
+      val existing = cache.putIfAbsent(key, missed)
+      if (existing ne null) existing else missed
+    }
+  }
+  override def optimise : Unit = {
+    //remove the default values
+    //cant use cache.values.remove(All), they only remove a single instance
+    val it = cache.entrySet().iterator()
+    while (it.hasNext()) {
+      val curr = it.next()
+      if (curr.getValue == defaultValue) it.remove()
+    }
+  }
+}
+//class CachedLazyIndexedMapping[V <: Named] (miss : (String => Iterable[V]) )
```

Comment:
it will be removed in the next push.
It was to get around the access of a Seq[File/class etc] by name
e.g. for findClass

replaced by an optimisation 
```  class PackageInfo(val packageName: String, val list: ClassPathEntries, val files: Seq[FileEntryType], val packages: Seq[PackageEntryImpl]) {
    lazy val filesByName: Map[String, AbstractFile] = files.map {
      file => file.name -> file.file
    }(collection.breakOut)
```